pkg/scheduler/scheduler.go

```diff
@@ -88,197 +47,231 @@ func NewScheduler(
 	metricsEmitter emitter.MetricsEmitter,
 	logger lager.Logger,
 ) *Scheduler {
+	retryInterval := defaultRetryInterval
+	if schedulerConfig.CollectorRetryIntervalMs != nil {
+		retryInterval = *schedulerConfig.CollectorRetryIntervalMs
+	}
+	maxRetries := defaultMaxRetries
+	if schedulerConfig.CollectorMaxRetries != nil {
+		maxRetries = *schedulerConfig.CollectorMaxRetries
+	}
 
 	return &Scheduler{
 		brokerinfo:     brokerInfo,
 		metricsEmitter: metricsEmitter,
 
 		instanceRefreshInterval: schedulerConfig.InstanceRefreshInterval,
+		collectorRetryInterval:  retryInterval,
+		collectorMaxRetries:     maxRetries,
 
 		metricsCollectorDrivers: map[string]collector.MetricsCollectorDriver{},
-		workers:                 collectorWorkerMap{workers: map[workerID]*collectorWorker{}},
+		workers:                 map[workerID]*collectorWorker{},
 
 		logger: logger,
 	}
 }
 
 // WithDriver ...
-func (w *Scheduler) WithDriver(drivers ...collector.MetricsCollectorDriver) *Scheduler {
+func (s *Scheduler) WithDriver(drivers ...collector.MetricsCollectorDriver) *Scheduler {
 	for _, driver := range drivers {
-		w.metricsCollectorDrivers[driver.GetName()] = driver
-		w.logger.Debug("registered_driver", lager.Data{"name": driver.GetName()})
+		s.metricsCollectorDrivers[driver.GetName()] = driver
+		s.logger.Debug("registered_driver", lager.Data{"name": driver.GetName()})
 	}
 
-	return w
+	return s
 }
 
-func (w *Scheduler) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
-	var err error
-	w.job, err = scheduler.Every(w.instanceRefreshInterval).Seconds().Run(func() {
-		w.workersStarting.Add(1)
-		defer w.workersStarting.Done()
+func (s *Scheduler) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
+	var ctx context.Context
+	ctx, s.cancel = context.WithCancel(context.Background())
+	close(ready)
```

Comment:
I'm not sure this synchronisation make sense here as we only create a context before we signal back.

Reply:
I have to put it somewhere, but before the `mainLoop` which is synchronous. I preferred to define the the variable first, and then initialise it, but I can swap the order. I can also pass the `ready` to `mainLoop` and then close it there, but it would almost be the same.